conn_js.go

```diff
@@ -0,0 +1,196 @@
+// This method is thread-safe.
+func (c *Conn) Close() error {
+	// c.mut.Lock()
+	// defer c.mut.Unlock()
+	c.Call("close")
+	if c.messageHandler != nil {
+		c.Call("removeEventListener", "message", *c.messageHandler)
+		c.messageHandler.Release()
+	}
+	if c.closeHandler != nil {
+		c.Call("removeEventListener", "close", *c.closeHandler)
+		c.closeHandler.Release()
+	}
+	return nil
+}
+
+func (c *Conn) LocalAddr() net.Addr {
+	// TODO(albrow): is there some way to get the local address?
+	return NewAddr("0.0.0.0:0")
+}
+
+func (c *Conn) RemoteAddr() net.Addr {
+	rawURL := c.Get("url").String()
+	withoutPrefix := strings.TrimPrefix(rawURL, "ws://")
+	withoutSuffix := strings.TrimSuffix(withoutPrefix, "/")
+	return NewAddr(withoutSuffix)
+}
+
+func (c *Conn) SetDeadline(t time.Time) error {
+	// Not supported
+	return nil
+}
+
+func (c *Conn) SetReadDeadline(t time.Time) error {
+	// Not supported
+	return nil
+}
+
+func (c *Conn) SetWriteDeadline(t time.Time) error {
+	// Not supported
+	return nil
+}
+
+// NewConn creates a Conn given a regular js/wasm WebSocket Conn.
+func NewConn(raw js.Value) *Conn {
+	conn := &Conn{
+		Value:        raw,
+		incomingData: make(chan []byte),
+		closeSignal:  make(chan struct{}),
+	}
+	conn.setUpHandlers()
+	return conn
+}
+
+func (c *Conn) setUpHandlers() {
+	// c.mut.Lock()
+	// defer c.mut.Unlock()
+	if c.messageHandler != nil {
+		// Message handlers already created. Nothing to do.
+		return
+	}
+	messageHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
+			// TODO(albrow): Currently we assume data is of type Blob. Really we
+			// should check binaryType and then decode accordingly.
+			blob := args[0].Get("data")
+			text := readBlob(blob)
+			c.incomingData <- []byte(text)
+		}()
+		return nil
+	})
+	c.messageHandler = &messageHandler
+	c.Call("addEventListener", "message", messageHandler)
+
+	closeHandler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		close(c.closeSignal)
+		return nil
+	})
+	c.closeHandler = &closeHandler
+	c.Call("addEventListener", "close", closeHandler)
+}
+
+func (c *Conn) waitForOpen() error {
+	openSignal := make(chan struct{})
+	handler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		close(openSignal)
+		return nil
+	})
+	defer func() {
+		c.Call("removeEventListener", "open", handler)
+	}()
+	defer handler.Release()
+	c.Call("addEventListener", "open", handler)
+	<-openSignal
+	return nil
+}
+
+func readBlob(blob js.Value) string {
+	// const reader = new FileReader();
+	// reader.addEventListener('loadend', (e) => {
+	//   const text = e.srcElement.result;
+	//   console.log(text);
+	// });
+	// reader.readAsText(blb);
+	reader := js.Global().Get("FileReader").New()
+	textChan := make(chan string)
+	loadEndFunc := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
+		go func() {
+			text := args[0].Get("srcElement").Get("result").String()
+			textChan <- text
```

Comment:
`textChan` should just have a buffer of one instead of launching a goroutine.

Reply:
One quirk of the `syscall/js` package is that you can't block inside of any JavaScript callback (i.e. anything constructed with `js.FuncOf`). Blocking inside a JavaScript callback blocks the entire JavaScript event loop and no other callbacks can be processed.

The latest version of this PR uses a buffer of 100 and a separate goroutine whose sole responsibility is reading from the channel. It is still technically possible to block inside the JavaScript callback but that would only happen if the buffered channel were full, which is incredibly unlikely. Let me know what you think of this approach.

Reply:
Here, a buffer of one would have been fine.

> Let me know what you think of this approach.

I think we should use a lock, buffer, and a tight critical section where we write into that buffer.